scanner/services/converters/converters.go

```diff
@@ -0,0 +1,400 @@
+}
+
+func toClairCorePackage(p *v4.Package) (string, *claircore.Package, error) {
+	// Conversion functions.
+	toNormalizedVersion := func(v *v4.NormalizedVersion) (ccV claircore.Version) {
+		ccV.Kind = v.GetKind()
+		copy(ccV.V[:], v.GetV())
+		return
+	}
+	toSourcePackage := func(p *v4.Package) (*claircore.Package, error) {
+		if p == nil {
+			return nil, nil
+		}
+		// Prevent deeper recursion.
+		if p.GetSource() != nil {
+			p.Source = nil
+		}
+		_, ccP, err := toClairCorePackage(p)
+		return ccP, err
+	}
+	// Fields that might fail.
+	ccCPE, err := toClairCoreCPE(p.GetCpe())
+	if err != nil {
+		return "", nil, fmt.Errorf("package %q: %w", p.GetId(), err)
+	}
+	src, err := toSourcePackage(p.GetSource())
+	if err != nil {
+		return "", nil, err
+	}
+	return p.GetId(), &claircore.Package{
+		ID:                p.GetId(),
+		Name:              p.GetName(),
+		Version:           p.GetVersion(),
+		Kind:              p.GetKind(),
+		Source:            src,
+		PackageDB:         p.GetPackageDb(),
+		RepositoryHint:    p.GetRepositoryHint(),
+		NormalizedVersion: toNormalizedVersion(p.GetNormalizedVersion()),
+		Module:            p.GetModule(),
+		Arch:              p.GetArch(),
+		CPE:               ccCPE,
+	}, nil
+}
+
+func toClairCoreDistributions(d *v4.Distribution) (string, *claircore.Distribution, error) {
+	ccCPE, err := toClairCoreCPE(d.GetCpe())
+	if err != nil {
+		return "", nil, fmt.Errorf("distribution %q: %w", d.GetId(), err)
+	}
+	return d.GetId(), &claircore.Distribution{
+		ID:              d.GetId(),
+		DID:             d.GetDid(),
+		Name:            d.GetName(),
+		Version:         d.GetVersion(),
+		VersionCodeName: d.GetVersionCodeName(),
+		VersionID:       d.GetVersionId(),
+		Arch:            d.GetArch(),
+		CPE:             ccCPE,
+		PrettyName:      d.GetPrettyName(),
+	}, nil
+}
+
+func toClairCoreRepository(r *v4.Repository) (string, *claircore.Repository, error) {
+	ccCPE, err := toClairCoreCPE(r.GetCpe())
+	if err != nil {
+		return "", nil, fmt.Errorf("repository %q: %w", r.GetId(), err)
+	}
+	return r.GetId(), &claircore.Repository{
+		ID:   r.Id,
+		Name: r.Name,
+		Key:  r.Key,
+		URI:  r.Uri,
+		CPE:  ccCPE,
+	}, nil
+}
+
+func toClairCoreEnvironment(env *v4.Environment) (*claircore.Environment, error) {
+	introducedIn, err := claircore.ParseDigest(env.GetIntroducedIn())
+	if err != nil {
+		return nil, err
+	}
+	return &claircore.Environment{
+		PackageDB:      env.GetPackageDb(),
+		IntroducedIn:   introducedIn,
+		DistributionID: env.GetDistributionId(),
+		RepositoryIDs:  env.GetRepositoryIds(),
+	}, nil
+}
+
+// convertToClairCoreMap converts a slice of pointers of a generic type to a map
+// based on the returned value of a conversion function that returns a string
+// key, the pointer to the converted value, or error if the conversion failed.
+// Nils in the slice are ignored.
+func convertToClairCoreMap[IN any, OUT any](in []*IN, convF func(*IN) (string, *OUT, error)) (map[string]*OUT, error) {
```

Comment:
nit: can we give this a more generic name like `convertToMapFunc`? The function doesn't involve ClairCore (just used for that purpose)